### DISCUSSION TITLE
Add search functionality to media

### DISCUSSION THREAD

**Discussion by OnFreund** [home-assistant/architecture#1038] [Entity Models]
Media sources are browsable, and media players can opt-in to be browsable.
I propose we also add search abilities to both. The implementation would be platform specific - e.g. the Spotify media players can use its [search api](https://developer.spotify.com/documentation/web-api/reference/search), Radio Browser would use its api, a YouTube media source can search it, etc...

This would be extremely useful when combined with assist (asking assist to search for and play your favorite artist, etc...)

Opting in to search means implementing:
```Python
async def async_browse_media(
  self,
  query: str,
  allowed_types: list[MediaClass],
  num_results: int,
  offset: int
)
```

Open questions:
* Can media sources be searchable but not browseable?
* Sorting and filtering?
* Displaying the results - grouped by class?

**Comment by allenporter** [comment]
There's a previous proposal here https://github.com/home-assistant/architecture/discussions/677 which had some positive support but was not implemented. Perhaps it is worth reviving, confirming and having the core team approve.

I would suggest not also adding new concepts for paging/offsets/results in the same proposal since its a concern that can be considered independent. That is, any media source browse call already has to handle limits or lack of limits. (If we *were* to introduce paging, then string resume token that the integration can encode would be preferred over integer offsets which are limiting)

**Reply by OnFreund** [reply]
Thanks, somehow missed that.

**Comment by Misiu** [comment]
There is a [frontend PR](https://github.com/home-assistant/frontend/pull/19199) that adds the desired functionality to the media browser, but I agree that this should be platform-specific to be able to use the API.
Right now it is so hard to pick radio when creating automation.

I'd also would like to see a favorite functionality. The radio browser uses external API, but I'd like to add radios to favorites to show only that list in the front end for easy picking radios via the interface.

**Reply by OnFreund** [reply]
This isn't exactly the same thing - the browser can only search what's available to it. It can't search the entire YouTube or Spotify catalog, for example.
